pkg/scheduler/partition.go

```diff
@@ -591,14 +590,12 @@ func (pc *PartitionContext) updatePartitionResource(delta *resources.Resource) {
 // This locks the partition. The partition may not be locked when we process the allocation
 // additions to the node as that takes further app, queue or node locks
 func (pc *PartitionContext) addNodeToList(node *objects.Node) error {
-	pc.Lock()
-	defer pc.Unlock()
-	// Node can be added to the system to allow processing of the allocations
 	if err := pc.nodes.AddNode(node); err != nil {
 		return fmt.Errorf("failed to add node %s to partition %s, error: %v", node.NodeID, pc.Name, err)
 	}
+	pc.Lock()
+	defer pc.Unlock()
```

Comment:
We postponed the partition lock to this point. 

The reason is that we don't want any go routine holding partition lock before changing  pc.nodes?

Reply:
I just ran a test for this PR but move the pc.Lock() back to the first line of addNodeToList(). 

There is no POTENTIAL DEADLOCK called "Inconsistent locking". Looks like this change is unnecessary.

Reply:
It's possible, however, we'd like to be consistent. If we decide not to grab the PartitionContext lock in other cases, it's better to do the same here, becase we avoid locks that are overlapping.

Reply:
Oh, I think I misunderstood.
This is an improvement to decouple locks of pc.nodes and partition.

It's not a change to fix deadlock issue.

Reply:
"decouple" -> I mean remove overlapping between pc.nodes and partition.

It's not quite straightforward to me when I see the lock postponed here, maybe we can add a comment here?

Reply:
Actually we can move it back. 

1) We already do something similar in `updatePartitionDetails()`. In `updateNodeSortingPolicy()` we make reference to `pc.nodes` while having a partition lock.
2) This can give an inconsistent view: node is already in the data structure, however, partition information (`totalPartitionResource`, `root.maxResource`) has been not updated yet. I don't think it's critical though, but it can trigger the health checker in the wrong moment.

The most important thing is NOT to have partition lock when calling back from the Application. We already achieved this task.

I'll let @wilfred-s share his opinion on this.

Reply:
To make it clear why don't we follow the pattern of the node removal?
Split the function in two:
* new function `addNodeResources(node)` that locks the partition and does everything after the metrics update.
* existing function `addNodeToList(node)` that does everything including the metrics update and then calls `addNodeResources(node)`

Add proper comments to the functions and it is not up for debate etc.
BTW: the small delay between the adding of the node and the update of the resource if it exists is not an issue.

Reply:
+1 for below 2 options
1. Spliting addNodeToList() into two function as Wilfred suggested
2. Add a comment to explain why we postponed the pc.Lock

+0 for below 1 option
1. moving pc.Lock back to first line

I agree it has become a matter of preference. I prefer the idea to remove lock overlapping. Add a comment here might be the simplest option.

All options are OK to me.